src/ethereum_spec_tools/sync.py

```diff
@@ -321,12 +414,36 @@ def fetch_blocks(
                     raise Exception("mismatched request id")
 
                 if "error" in reply:
-                    blocks[reply_id] = RpcError(
+                    raise RpcError(
                         reply["error"]["code"],
                         reply["error"]["message"],
                     )
                 else:
-                    blocks[reply_id] = bytes.fromhex(reply["result"])
+                    res = reply["result"]
+                    header = self.make_header(res)
+                    transactions = []
+                    for t in res["transactions"]:
+                        transactions.append(
+                            self.module("eth_types").Transaction(
+                                hex_to_u256(t["nonce"]),
+                                hex_to_u256(t["gasPrice"]),
+                                hex_to_u256(t["gas"]),
+                                self.module(
+                                    "utils.hexadecimal"
+                                ).hex_to_address(t["to"])
+                                if t["to"]
+                                else Bytes0(b""),
+                                hex_to_u256(t["value"]),
+                                hex_to_bytes(t["input"]),
+                                hex_to_u256(t["v"]),
+                                hex_to_u256(t["r"]),
+                                hex_to_u256(t["s"]),
+                            )
+                        )
+                    uncles = self.fetch_uncles(reply_id, len(res["uncles"]))
```

Comment:
Oof, so now we have one extra request per block? Any way we can collect all the uncle fetches into one giant request, so we only have two requests (one to get blocks, and one to get all their uncles)?

Reply:
One extra call for *every uncle* in each block :D